src/main/java/org/truffleruby/core/hash/library/PackedHashStoreLibrary.java

```diff
@@ -154,6 +158,16 @@ private static void promoteToBuckets(RubyHash hash, Object[] store, int size) {
         hash.size = size;
     }
 
+    @TruffleBoundary
+    private static void promoteToCompact(RubyHash hash, Object[] store, int size) {
+        CompactHashStore newStore = new CompactHashStore(size);
```

Comment:
I think it would be nicer here to pass `PROMOTION_SIZE` and define that above as `PROMOTION_SIZE = MAX_ENTRIES + 1`, that likely compiles slightly better.
And we can `assert size == PROMOTION_SIZE` before  `promoteToBigHash.execute(node, store, hash);` below.

Reply:
In more details it's valuable that the compiler knows a fixed compile-time array size, because then it can use partial escape analysis on the arrays and otherwise it cannot.